### PR TITLE
Add the default eslint react hooks config to the wordpress eslint package

### LINKED ISSUE #14676: Add the React hooks eslint to our default eslint config
Also check if can avoid conflicts with the late variable declaration rule.

See here https://github.com/WordPress/gutenberg/pull/14612#discussion_r268909604

cc @aduth

### PR DESCRIPTION
**Pull request by youknowriad**
closes #14676

This PR adds the default eslint config for react hooks (See https://reactjs.org/docs/hooks-rules.html) to the WordPress eslint config.

 - We were already violating our rules in the contrast checker component.
 - It's also interesting to note that we are currently violating a "recommendation" in the `URLPopoverAtLink` component. It's our use-case is legitimate though so I was wondering if we should remove the second rule (the warning) added here. cc @aduth

### CODE REVIEW

**Review by ntwb — APPROVED** [review]
LGTM 👍

That said, what's the status of changelogs at the moment, does this PR require it?


Edit: If so please add it, if not merge away 🎉

**Review by gziolo — APPROVED** [review]
Nice 👍

**Comment by youknowriad** [comment]
@ntwb I already have a CHANGELOG entry

File: packages/block-editor/src/components/contrast-checker/index.js [hunk 1]
```diff
@@ -11,6 +11,22 @@ import { __ } from '@wordpress/i18n';
 import { Notice } from '@wordpress/components';
 import { useEffect } from '@wordpress/element';
 
+function ContrastCheckerMessage( { tinyBackgroundColor, tinyTextColor, backgroundColor, textColor } ) {
```
Comment (gziolo) [inline packages/block-editor/src/components/contrast-checker/index.js]: Out of curiosity, what ESLint rule was triggered which enforced this refactor? I'm wondering what is the lesson learned.

Comment (youknowriad) [reply packages/block-editor/src/components/contrast-checker/index.js]: The reason is that you can use hooks after "early returns"

File: packages/eslint-plugin/configs/react.js [hunk 2]
```diff
@@ -24,5 +25,7 @@ module.exports = {
 		'react/no-children-prop': 'off',
 		'react/prop-types': 'off',
 		'react/react-in-jsx-scope': 'off',
+		'react-hooks/rules-of-hooks': 'error',
```
Comment (gziolo) [inline packages/eslint-plugin/configs/react.js]: Interesting, that they don't provide the recommended preset.

File: packages/eslint-plugin/configs/react.js [hunk 3]
```diff
@@ -24,5 +25,7 @@ module.exports = {
...
+		'react-hooks/exhaustive-deps': 'warn',
```
Comment (youknowriad) [inline packages/eslint-plugin/configs/react.js]: I'm hesitant on the need of this one as we already have a valid use-case for extra deps, but I'm merging and we can try and see if it feels constraining.

File: packages/block-editor/src/components/contrast-checker/test/__snapshots__/index.js.snap [hunk 4]
```diff
@@ -8,24 +8,57 @@ exports[`ContrastChecker should render component when the colors do not meet AA
...
+        "_a": 1,
+        "_b": 102,
+        "_format": "hex",
+        "_g": 102,
+        "_gradientType": undefined,
+        "_ok": true,
+        "_originalInput": "#666",
+        "_r": 102,
+        "_roundA": 1,
+        "_tc_id": 2,
```
Comment (aduth) [inline packages/block-editor/src/components/contrast-checker/test/__snapshots__/index.js.snap]: These IDs increment with each use, thus can be affected externally and fail the tests.

See: 4c6a9a4 / #18681

Do we have to pass around the TinyColor object? Should we avoid snapshots?

Comment (youknowriad) [reply packages/block-editor/src/components/contrast-checker/test/__snapshots__/index.js.snap]: yes, these snapshots are not that useful I think.

Comment (aduth) [reply packages/block-editor/src/components/contrast-checker/test/__snapshots__/index.js.snap]: See #19169

**Pull request merged**